test/unit/utils.test.js

```diff
@@ -297,4 +299,66 @@ describe('utils', function () {
       });
     });
   });
+
+  describe('shuffler function', () => {
+    it('should support iterables', function () {
+      // Kind of an implicit test, we should not throw/crash here.
+      const input = new Set(['a', 'b', 'c']);
+      const output = shuffle(input);
+      expect(Array.isArray(output)).to.be.true;
+    });
+
+    it('should not mutate the original input', function () {
+      const input = Object.freeze(['a', 'b', 'c', 'd', 'e']);
+      const output = shuffle(input); // This will throw if shuffle tries to edit the input
+      expect(output).to.not.deep.equal(input);
+      expect(output).to.have.lengthOf(input.length);
+    });
+
+    it('should give a random subset if limit is less than the input length', () => {
+      const input = ['a', 'b', 'c', 'd', 'e'];
+      const output = shuffle(input, 3);
+      expect(output).to.have.lengthOf(3);
+    });
+
+    const input = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i'];
+    for (let n = 1; n <= input.length; n++) {
+      it(`should give a random subset of length ${n}`, function () {
```

Comment:
I appreciate the idea here, but the resulting tests are going to read as "should give a random subset of length 5", for example, which doesn't clearly convey the intent; the important things to test for any given input are the boundaries (where the behavior changes): default, `0`, `1`, `n-1`, and `n`. We have default and 0 covered, so, I recommend taking the logic here using it to replace the test case in L318, and just test the `input.length - 1`. To cover `1`, we can actually have two test cases, one where the input is only 1 element, to make sure we return that element, and one where the input is greater (feel free to keep using the 9 element array), to prevent excessive flakiness, run the shuffle 5 times with limit 1 and then make sure that the set of all returned elements has size > 1. And then reuse a copy of the L337 test case to test `limit === input.length`.